generative_ai/prompts/test_prompt_optimizer.py

```diff
@@ -0,0 +1,143 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+import random
+import time
+from typing import Callable
+
+from google.cloud import aiplatform, storage
+from google.cloud.aiplatform import CustomJob
+from google.cloud.aiplatform_v1 import JobState
+from google.cloud.exceptions import NotFound
+from google.cloud.storage import transfer_manager
+
+from prompt_optimizer import optimize_prompts
+
+import pytest
+
+PROJECT_ID = os.getenv("GOOGLE_CLOUD_PROJECT")
+STAGING_BUCKET_NAME = "prompt_optimizer"
+CONFIG_SOURCE_DIRECTORY = "config"
+CONFIGURATION_FILENAME = "sample_configuration.json"
+LOCATION = random.choice(["us-central1", "us-east4", "us-west4", "us-west1"])
+OUTPUT_PATH = "instruction"
+
+STORAGE_CLIENT = storage.Client()
+
+
+def _clean_resources(bucket_resource_name: str) -> None:
+    # delete blobs and bucket if exists
+    try:
+        bucket = STORAGE_CLIENT.get_bucket(bucket_resource_name)
+    except NotFound:
+        print(f"Bucket {bucket_resource_name} cannot be accessed")
+        return
+
+    blobs = bucket.list_blobs()
+    for blob in blobs:
+        blob.delete()
+    bucket.delete()
+
+
+def substitute_env_variable(data: dict, target_key: str, env_var_name: str) -> dict:
+    # substitute env variables in the given config file with runtime values
+    if isinstance(data, dict):
+        for key, value in data.items():
+            if key == target_key:
+                data[key] = os.environ.get(env_var_name)
+            else:
+                data[key] = substitute_env_variable(value, target_key, env_var_name)
+    elif isinstance(data, list):
+        for i, value in enumerate(data):
+            data[i] = substitute_env_variable(value, target_key, env_var_name)
+    return data
+
+
+def update_json() -> dict:
+    # Load the JSON file
+    with open(f"/{CONFIG_SOURCE_DIRECTORY}/{CONFIGURATION_FILENAME}", "r") as f:
+        data = json.load(f)
+    # Substitute only the "project" variable with the value of "PROJECT_ID"
+    substituted_data = substitute_env_variable(data, "project", "PROJECT_ID")
+    return substituted_data
+
+
+@pytest.fixture(scope="session")
+def bucket_name() -> str:
+    filenames = [
+        "sample_prompt_template.txt",
+        "sample_prompts.jsonl",
+        "sample_system_instruction.txt",
+    ]
+    # cleanup existing stale resources
+    _clean_resources(STAGING_BUCKET_NAME)
+    # create bucket
+    bucket = STORAGE_CLIENT.bucket(STAGING_BUCKET_NAME)
+    bucket.storage_class = "STANDARD"
+    new_bucket = STORAGE_CLIENT.create_bucket(bucket, location="us")
```

Comment:
suggestion: Please consider creating the bucket only if it doesn't exist. With current approach we risk that someone will grab our bucket name (name must be globally unique) and the test will start failling.

Reply:
Great catch!

I was weighing against having the bucket as a permanent resource vs creating it just for the test. Agree that there could potentially be a name clash even if its implemented dynamically.

Looks like the better approach is to upload static config files into a bucket in all the CI projects? If so, kindly help with provisioning.

Reply:
SG, please sync with Sampath on this. Nevertheless I would still keep the files in the source code as a reference for people who download the sample/clone the repo.